File: golang/algorithms/turbo-reverse-factor.go
package algorithms

// TurboReverseFactor tìm kiếm chuỗi con bằng thuật toán Turbo Reverse Factor
// Độ phức tạp thời gian: O(n+m) trong trường hợp xấu nhất, O(n+m) trong trường hợp tốt nhất
// Độ phức tạp không gian: O(m) cho bảng shift
// Trong đó n là độ dài của chuỗi source, m là độ dài của chuỗi target
//
// Turbo Reverse Factor là cải tiến của Reverse Factor
// So sánh từ phải sang trái, sử dụng turbo optimization để tránh so sánh lại các ký tự đã khớp
func TurboReverseFactor(source, target string) (exist bool, index int) {
	exist = false
	index = -1

	sourceLen := len(source)
	targetLen := len(target)

	// Kiểm tra trường hợp đặc biệt
	if targetLen == 0 {
		return true, 0
	}

	if targetLen > sourceLen {
		return false, -1
	}

	// Tính failure function (giống KMP)
	failure := make([]int, targetLen)
	failure[0] = 0
	j := 0

	for i := 1; i < targetLen; i++ {
		for j > 0 && target[i] != target[j] {
			j = failure[j-1]
		}
		if target[i] == target[j] {
			j++
		}
		failure[i] = j
	}

	// Tính factor table
	factor := make([]int, targetLen+1)
	factor[0] = 1
	j = 0

	for i := 1; i <= targetLen; i++ {
		for j > 0 && target[i-1] != target[j] {
			j = failure[j-1]
		}
		if target[i-1] == target[j] {
			j++
		}
		factor[i] = i - j
	}

	// Bắt đầu tìm kiếm
	i := 0

	for i <= sourceLen-targetLen {
		j = targetLen - 1

		// So sánh từ phải sang trái
		// Turbo optimization: bỏ qua các ký tự đã khớp từ lần trước
		for j >= 0 && source[i+j] == target[j] {
			j--
		}

		// Nếu j < 0, tìm thấy match
		if j < 0 {
			exist = true
			index = i
			return exist, index
		}

		// Tính shift sử dụng factor table
		shift := factor[j+1]
		if shift == 0 {
			shift = 1
		}
		i += shift
	}

	return exist, index
}

File: nodejs/algorithms/turboReverseFactor.js
/**
 * TurboReverseFactor - Thuật toán tìm kiếm chuỗi con bằng Turbo Reverse Factor
 * 
 * Độ phức tạp thời gian: O(n+m)
 * Độ phức tạp không gian: O(m)
 * 
 * Turbo Reverse Factor là cải tiến của Reverse Factor
 * So sánh từ phải sang trái, sử dụng turbo optimization để tránh so sánh lại các ký tự đã khớp
 */

class TurboReverseFactor {
    static search(source, target) {
        let exist = false;
        let index = -1;

        const sourceLen = source.length;
        const targetLen = target.length;

        if (targetLen === 0) {
            return { exist: true, index: 0 };
        }

        if (targetLen > sourceLen) {
            return { exist: false, index: -1 };
        }

        // Tính failure function (giống KMP)
        const failure = new Array(targetLen).fill(0);
        failure[0] = 0;
        let j = 0;

        for (let i = 1; i < targetLen; i++) {
            while (j > 0 && target[i] !== target[j]) {
                j = failure[j - 1];
            }
            if (target[i] === target[j]) {
                j++;
            }
            failure[i] = j;
        }

        // Tính factor table
        const factor = new Array(targetLen + 1).fill(0);
        factor[0] = 1;
        j = 0;

        for (let i = 1; i <= targetLen; i++) {
            while (j >= 0 && (j === targetLen || target[i - 1] !== target[j])) {
                factor[i] = i - j;
                j = failure[j];
                if (j < 0) {
                    j = 0;
                }
            }
            j++;
        }

        // Bắt đầu tìm kiếm
        let i = 0;
        let turboShift = 0;

        while (i <= sourceLen - targetLen) {
            j = targetLen - 1;

            // So sánh từ phải sang trái
            // Turbo optimization: bỏ qua các ký tự đã khớp từ lần trước
            while (j >= 0 && j >= turboShift && source[i + j] === target[j]) {
                j--;
            }

            // Nếu j < turboShift, tìm thấy match
            if (j < turboShift) {
                exist = true;
                index = i;
                return { exist, index };
            }

            // Nếu j < 0, tìm thấy match
            if (j < 0) {
                exist = true;
                index = i;
                return { exist, index };
            }

            // Tính shift sử dụng factor table
            const shift = factor[j + 1];
            turboShift = targetLen - shift;
            i += shift;
        }

        return { exist, index };
    }

    static searchAll(source, target) {
        const positions = [];
        const sourceLen = source.length;
        const targetLen = target.length;

        if (targetLen === 0 || targetLen > sourceLen) {
            return positions;
        }

        let offset = 0;
        while (offset <= sourceLen - targetLen) {
            const substring = source.substring(offset);
            const result = this.search(substring, target);

            if (result.exist) {
                positions.push(offset + result.index);
                offset += result.index + 1;
            } else {
                break;
            }
        }

        return positions;
    }

    static count(source, target) {
        return this.searchAll(source, target).length;
    }
}

module.exports = TurboReverseFactor;

